snuba/datasets/schemas/join.py

```diff
@@ -0,0 +1,126 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import NamedTuple, Optional, Sequence
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.datasets.schemas import Schema
+
+
+class JoinType(Enum):
+    INNER = "INNER"
+    LEFT = "LEFT"
+    RIGHT = "RIGHT"
+    FULL = "FULL"
+
+
+class JoinConditionExpression(NamedTuple):
+    """
+    Represent one qualified column [alias.column] in the
+    ON clause within the join expression.
+    """
+    table_alias: str
+    column: str
+
+
+@dataclass(frozen=True)
+class JoinCondition:
+    """
+    Represent a condition in the ON clause in the JOIN expression
+    """
+    left: JoinConditionExpression
+    right: JoinConditionExpression
+
+    def get_join_condition(self) -> str:
```

Comment:
On second thought, maybe it would make sense to change this and the `print() -> str` methods to `__str__() -> str` methods so you can just put them in the f-strings themselves? Usually, I try and avoid those special methods, but it seems in this case, these structures are literally intended for join templating, so it probably makes sense in this case. I also feel like the word `print` is a little overloaded in Python and might make sense to avoid below.